typescript/lib/parsers/dalle.ts

```diff
@@ -0,0 +1,852 @@
+import { JSONObject, JSONValue } from "../../common";
+import {
+  Prompt,
+  Output,
+  PromptInput,
+  ModelMetadata,
+  ExecuteResult,
+} from "../../types";
+import { AIConfigRuntime } from "../config";
+import { ParameterizedModelParser } from "../parameterizedModelParser";
+import OpenAI, { ClientOptions } from "openai";
+import { getAPIKeyFromEnv } from "../utils";
+import { encode as gpt3Encode, decode as gpt3Decode } from "gpt-3-encoder";
+import {
+  CompletionCreateParams,
+  Chat,
+  CompletionCreateParamsNonStreaming,
+  CompletionCreateParamsStreaming,
+} from "openai/resources";
+import _ from "lodash";
+import { InferenceOptions } from "../modelParser";
+import { CallbackEvent } from "../callback";
+
+export class OpenAIModelParser extends ParameterizedModelParser<CompletionCreateParams> {
+  private openai: OpenAI | null = null;
+  private openaiOptions: ClientOptions | undefined;
+
+  public constructor(options?: ClientOptions) {
+    super();
+    this.openaiOptions = options;
+  }
+
+  public serialize(
+    promptName: string,
+    data: CompletionCreateParams,
+    aiConfig: AIConfigRuntime,
+    params?: JSONObject
+  ): Prompt {
+    // Serialize prompt input
+    let input: PromptInput;
+    if (typeof data.prompt === "string") {
+      input = data.prompt;
+    } else {
+      input = {
+        data: data.prompt,
+      };
+    }
+
+    // Serialize model metadata
+    let modelMetadata: ModelMetadata | string;
+    const promptModelMetadata: JSONObject = { ...data };
+    // Remove the prompt from the model data since that is not part of the model settings
+    delete promptModelMetadata.prompt;
+
+    // Check if AIConfig already has the model settings in its metadata
+    const modelName = data.model ?? this.id;
+    const globalModelMetadata = aiConfig.metadata.models?.[modelName];
+
+    if (globalModelMetadata != null) {
+      // Check if the model settings from the input data are the same as the global model settings
+
+      // Compute the difference between the global model settings and the model settings from the input data
+      // If there is a difference, then we need to add the different model settings as overrides on the prompt's metadata
+      const keys = _.union(
+        _.keys(globalModelMetadata),
+        _.keys(promptModelMetadata)
+      );
+      const overrides = _.reduce(
+        keys,
+        (result: JSONObject, key) => {
+          if (!_.isEqual(globalModelMetadata[key], promptModelMetadata[key])) {
+            result[key] = promptModelMetadata[key];
+          }
+          return result;
+        },
+        {}
+      );
+
+      if (Object.keys(overrides).length > 0) {
+        modelMetadata = {
+          name: modelName,
+          settings: overrides,
+        };
+      } else {
+        modelMetadata = modelName;
+      }
+    } else {
+      modelMetadata = {
+        name: modelName,
+        settings: promptModelMetadata,
+      };
+    }
```

Comment:
The TS SDK [has a method `getModelMetadata()`](https://github.com/lastmile-ai/aiconfig/blob/main/typescript/lib/config.ts#L898)

Similar to what you did in the [python impl](https://github.com/lastmile-ai/aiconfig/pull/288/files#diff-ea23df90ffdfdbb7dc70b4283145953a1a62ecada5b8aa66d412575da271955dR118)